### PR TITLE
Remove join-pairs from join processing

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

The main concept is:

- Translate the tree structure from the joins in the query-ast into the tree structure of the logical plan without the intermediate transformation into join pairs.
- Translate an implicit join with join-condition in the where-clause into a cross-join with filter and then add a rule to merge the filter with the equi-join condition into the cross-join to an inner-join. 

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Review by mfussenegger — COMMENTED** [review]
I'm not sure if this moves in the right direction.

I kinda like the part of having a proper _per_ join relation, but my gut feeling is that the analyzer should preserve all information, not throw it away. We'd loose the association between the join condition and join.

So I think instead we should have a `JoinRelation` that preserves the rhs/lhs, the join type and join condition.

We could then still use the filter+optimization approach by adding a normalization step somewhere in the planner, if that simplifies the logic.

**Comment by mkleen** [comment]
> I kinda like the part of having a proper _per_ join relation, but my gut feeling is that the analyzer should preserve all information, not throw it away. We'd loose the association between the join condition and join.

So you would prefer that the analyzer already associates the join-condition from the where with the JoinRelation?

**Comment by mkleen** [comment]
> > I kinda like the part of having a proper _per_ join relation, but my gut feeling is that the analyzer should preserve all information, not throw it away. We'd loose the association between the join condition and join.
> 
> So you would prefer that the analyzer already associates the join-condition from the where clause with the JoinRelation?

Btw, for all other joins than implicit joins all information is kept.

**Comment by mfussenegger** [comment]
> > I kinda like the part of having a proper _per_ join relation, but my gut feeling is that the analyzer should preserve all information, not throw it away. We'd loose the association between the join condition and join.
> 
> So you would prefer that the analyzer already associates the join-condition from the where clause with the JoinRelation?

No, it shouldn't do transformations. It should keep the shape of the original statement - including all its information.

If the user wrote `SELECT * FROM t1 INNER JOIN t2 on t1.x = t2.x` then we should be able to output the same statement after analysis.

This is a bit related to issues like:

- https://github.com/crate/crate/pull/8692
- https://github.com/crate/crate/pull/14491

Where we had considered creating a SQLFormatter based on the analysis result, but the current join representation makes it impossible, because our processing is lossy.

**Comment by mkleen** [comment]
> No, it shouldn't do transformations. It should keep the shape of the original statement - including all its information.
> 
> If the user wrote `SELECT * FROM t1 INNER JOIN t2 on t1.x = t2.x` then we should be able to output the same statement after analysis.

So you don't like the nesting of the tables for the joins already on analyzer level ?

**Comment by mkleen** [comment]
So we should avoid the transformation into a join relation?

**Comment by mkleen** [comment]
I moved the implicit join transformation into the planner.

File: server/src/main/java/io/crate/planner/optimizer/rule/MoveConstantJoinConditionsBeneathNestedLoop.java [hunk 1]
```diff
@@ -36,33 +36,31 @@
...
 import io.crate.sql.tree.JoinType;
-import io.crate.planner.operators.HashJoin;
 import io.crate.planner.operators.LogicalPlan;
-import io.crate.planner.operators.NestedLoopJoin;
 import io.crate.planner.optimizer.Rule;
 import io.crate.planner.optimizer.matcher.Captures;
 import io.crate.planner.optimizer.matcher.Pattern;
 
-public class MoveConstantJoinConditionsBeneathNestedLoop implements Rule<NestedLoopJoin> {
+public class MoveConstantJoinConditionsBeneathNestedLoop implements Rule<JoinPlan> {
```
Comment (matriv) [inline server/src/main/java/io/crate/planner/optimizer/rule/MoveConstantJoinConditionsBeneathNestedLoop.java:47]:
```suggestion
public class MoveConstantJoinConditionsBeneathJoin implements Rule<JoinPlan> {
```

**Comment by mkleen** [comment]
retest this please

**Pull request closed without merging**